modules/apps/transfer/types/packet_test.go

```diff
@@ -747,6 +788,63 @@ func TestUnmarshalPacketData(t *testing.T) {
 			},
 			nil,
 		},
+		{
+			"success: v2 with JSON encoding",
+			func() {
+				packetData := types.NewFungibleTokenPacketDataV2(
+					[]types.Token{
+						{
+							Denom:  types.NewDenom("atom", types.NewHop("transfer", "channel-0")),
+							Amount: "1000",
+						},
+					}, sender, receiver, "", types.ForwardingPacketData{})
+
+				packetDataBz, _ = json.Marshal(packetData)
+				version = types.V2
+				encoding = types.EncodingJSON
+			},
+			nil,
+		},
+		{
+			"failure: v2 with ABI encoding (not supported)",
+			func() {
+				packetDataBz = []byte("should fail before this gets relevant")
+				version = types.V2
+				encoding = types.EncodingABI
+			},
+			errors.New("encoding application/x-solidity-abi is only supported for ICS20-V1"),
+		},
+		{
+			"failure: invalid encoding",
+			func() {
+				encoding = "invalid"
+			},
+			errors.New("invalid encoding provided"),
+		},
+		{
+			"failure: invalid type for json",
+			func() {
+				packetDataBz = []byte("invalid")
+				encoding = types.EncodingJSON
+			},
+			ibcerrors.ErrInvalidType,
+		},
+		{
+			"failure: invalid type for protobuf",
+			func() {
+				packetDataBz = []byte("invalid")
```

Comment:
[nit] maybe we can extract it somewhere since we use it in at least three places